File: web/static/js/api/profiles.js
import { state } from "../stores/state.js";
import { fetchJson } from "./client.js";

export async function loadBotProfiles() {
  const data = await fetchJson("/api/bot-profiles");
  state.botProfiles = data.profiles || [];
  return state.botProfiles;
}

export async function createBotProfile(profile) {
  const data = await fetchJson("/api/bot-profiles", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(profile),
  });
  state.botProfiles = data.profiles || state.botProfiles;
  return data.profile;
}

export async function updateBotProfile(id, profile) {
  const data = await fetchJson(`/api/bot-profiles/${encodeURIComponent(id)}`, {
    method: "PATCH",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(profile),
  });
  state.botProfiles = data.profiles || state.botProfiles;
  return data.profile;
}

export async function deleteBotProfile(id) {
  const data = await fetchJson(`/api/bot-profiles/${encodeURIComponent(id)}`, { method: "DELETE" });
  state.botProfiles = data.profiles || state.botProfiles;
  return data.ok;
}

File: backend/api/profiles.py
from __future__ import annotations

from fastapi import APIRouter, Body, HTTPException, Request

from api.dependencies import require_local_service_control


def create_profiles_router() -> APIRouter:
    router = APIRouter()

    @router.get("/api/bot-profiles")
    async def bot_profiles(request: Request):
        require_local_service_control(request)
        return request.app.state.bot_profiles.list_profiles()

    @router.post("/api/bot-profiles")
    async def create_bot_profile(request: Request, payload: dict | None = Body(default=None)):
        require_local_service_control(request)
        try:
            profile = request.app.state.bot_profiles.create(payload or {})
        except ValueError as exc:
            raise HTTPException(status_code=409, detail=str(exc)) from exc
        return {"profile": profile, **request.app.state.bot_profiles.list_profiles()}

    @router.patch("/api/bot-profiles/{profile_id}")
    async def update_bot_profile(profile_id: str, request: Request, payload: dict | None = Body(default=None)):
        require_local_service_control(request)
        try:
            profile = request.app.state.bot_profiles.update(profile_id, payload or {})
        except KeyError as exc:
            raise HTTPException(status_code=404, detail="Profile not found") from exc
        return {"profile": profile, **request.app.state.bot_profiles.list_profiles()}

    @router.delete("/api/bot-profiles/{profile_id}")
    async def delete_bot_profile(profile_id: str, request: Request):
        require_local_service_control(request)
        return {"ok": request.app.state.bot_profiles.delete(profile_id), **request.app.state.bot_profiles.list_profiles()}

    return router
